paperwork_backend/shell.py

```diff
@@ -424,6 +426,48 @@ def _get_importer(fileuris, doc):
         return importers[int(idx)]
 
 
+def _do_ocr(empty_only, ocr_lang, ocr, pages):
+    input_queue = Queue(maxsize=0)
+    output_queue = Queue(maxsize=0)
+    num_threads = os.cpu_count()
+
+    for page in pages:
+        if empty_only and len(page.boxes) > 0:
+            continue
+        input_queue.put(page)
+
+    for i in range(num_threads):
+        worker = Thread(
+                target=_do_ocr_thread,
+                args=(ocr_lang, ocr, input_queue, output_queue)
+                )
+        worker.setDaemon(True)
+        worker.start()
+
+    input_queue.join()
+    pages = []
+
+    while not output_queue.empty():
+        page = output_queue.get()
+        pages.append(page)
+        output_queue.task_done()
+
+    return pages
+
+
+def _do_ocr_thread(ocr_lang, ocr, input_queue, output_queue):
+    while not input_queue.empty():
+        page = input_queue.get()
```

Comment:
What will happen if:
1) we have only 1 page to OCR
2) 2 threads (or more) are started
2) 2 threads validate the condition `not input_queue.empty()` at the very same time
3) and they both run `input_queue.get()`
?

2 ways to fix that:
1) Add a lock here preventing 2 threads from running `empty()` + `get()` both at the same time.
2) Just call `get()`:
```py
try:
    while True:
        page = input_queue.get(block=False)
        (...)
except queue.Empty:
    pass
````

Reply:
That is what I meant with 
> Further coding for `multiprocessing` is in my opinion simpler than for `threading`.

;)

I'll have a look later.